metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala

```diff
@@ -13,21 +13,24 @@ import java.nio.file.Paths
 
 trait ConfDecoder[A] { self =>
 
+  type T = A
   def read(conf: Conf): Configured[A]
+  def read(conf: Conf, state: => A): Configured[A]
+
   final def read(conf: Configured[Conf]): Configured[A] =
     conf.andThen(self.read)
 
   final def map[B](f: A => B): ConfDecoder[B] =
-    self.flatMap(x => Ok(f(x)))
+    new ConfDecoder.WithoutState[B] {
```

Comment:
I think in the direction of such API: what if user will call `map/flatMap` and other combinators on `ConfDecoderFactory` which will propagate state over transformations chain? And at the it he calls `.decoder` to generate final `ConfDecoder` with all context.

Reply:
the problem was that `state: => B` cannot be passed to underlying `ConfDecoder[A]`, and not to break compatibility, I added `mapExt` with covariance and `in/out` (maybe an implicit `CanBuildFrom` would also work).

read about type classes, very interesting (however, they don't require a single method and even allow inheritance):
* https://scalac.io/typeclasses-in-scala/ (also https://github.com/mpilquist/simulacrum)
* https://typelevel.org/cats/typeclasses.html
* https://alvinalexander.com/scala/fp-book/type-classes-101-introduction/

Reply:
My primary motivation to keep a single abstract method is that it makes it easier to reason about the meaning of methods like `map()`. It's the same reason I want to avoid multi-layer typeclasses because implicit resolution gets hairy when you have inheritance and you want to influence the priority of how typeclass instances get selected.